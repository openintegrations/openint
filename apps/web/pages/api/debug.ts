import type {NextApiHandler} from 'next'

export default ((_req, res) => {
  console.log('req.query', _req.query)
  res.send({ok: true})
}) satisfies NextApiHandler
