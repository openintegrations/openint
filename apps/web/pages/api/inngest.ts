import {serve} from 'inngest/next'
import type {Events} from '@openint/engine-backend/events'
import {inngest, outgoingWebhookEventMap} from '@openint/engine-backend/events'
import {withLog} from '@openint/util'
import * as functions from '../../inngest/functions'
import {sendWebhook} from '../../inngest/routines'

export const config = {
  maxDuration: 5 * 60, // 5 mins
}

export default serve(
  withLog('Starting inngest with', {
    client: inngest,
    functions: [
      ...Object.values(functions),
      // MARK: - Workaround for Inngest not having support for
      // multiple event triggers in a single function
      // @see https://discord.com/channels/842170679536517141/1214066130860118087/1214283616327180318
      ...Object.keys(outgoingWebhookEventMap).map((name) =>
        inngest.createFunction(
          {id: `send-webhook/${name}`},
          {event: name as keyof Events},
          sendWebhook,
        ),
      ),
    ],

    // landingPage: process.env['VERCEL_ENV'] !== 'production',
    logLevel: 'warn',
    // Enforce dev env never hit production inngest
    // https://discord.com/channels/842170679536517141/1080275520861782096/1080494988741324870
    baseUrl:
      // For debugging...
      process.env['INNGEST_REGISTER_URL'] ??
      (process.env.NODE_ENV === 'development'
        ? 'http://localhost:8288/fn/register'
        : undefined),
    // For debugging, via e.g. 'http://localhost:3010' via mitmproxy
    serveHost: process.env['INNGEST_SERVE_HOST'],
  }),
)
