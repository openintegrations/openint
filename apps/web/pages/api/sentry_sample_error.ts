import type {NextApiHandler} from 'next'

import {serverAnalytics} from '../../lib-server/analytics-server'

//  true
export default (async (_req, _res) => {
  serverAnalytics.track('test', {name: 'debug/debug', data: {}})
  await serverAnalytics.flush()
  throw new Error('Sentry Backend Error')
}) satisfies NextApiHandler
