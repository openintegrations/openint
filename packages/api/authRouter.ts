import {clerkClient} from '@clerk/nextjs/server'
import {z} from '@opensdks/util-zod'
import {TRPCError} from '@trpc/server'
import type {Viewer} from '@openint/cdk'
import {zViewer} from '@openint/cdk'
import {zOrganization, zUser} from '@openint/engine-backend'
import {
  adminProcedure,
  publicProcedure,
  trpc,
} from '@openint/engine-backend/router/_base'
import {R} from '@openint/util'

export type ClerkOrg = Awaited<
  ReturnType<(typeof clerkClient)['organizations']['getOrganization']>
>

export type ClerkUser = Awaited<
  ReturnType<(typeof clerkClient)['users']['getUser']>
>

export const authRouter = trpc.router({
  getViewer: publicProcedure
    .meta({
      openapi: {
        method: 'GET',
        path: '/viewer',
        summary: 'Get current viewer accessing the API',
      },
    })
    .input(z.void())
    .output(zViewer)
    .query(async ({ctx}) => {
      const extra =
        ctx.viewer.role === 'org'
          ? await clerkClient.organizations.getOrganization({
              organizationId: ctx.viewer.orgId,
            })
          : ctx.viewer.role === 'user'
            ? await clerkClient.users.getUser(ctx.viewer.userId)
            : undefined

      return {...ctx.viewer, extra} as Viewer
    }),

  getCurrentOrganization: adminProcedure
    .meta({
      openapi: {
        method: 'GET',
        path: '/viewer/organization',
        summary: 'Get current organization of viewer accessing the API',
      },
    })
    .input(z.void())
    .output(zOrganization.omit({privateMetadata: true}))
    .query(async ({ctx}) => {
      if (!ctx.viewer.orgId) {
        throw new TRPCError({code: 'BAD_REQUEST', message: 'orgId needed'})
      }
      return await getOrganizationOmitPrivateMeta(ctx.viewer.orgId)
    }),

  updateCurrentOrganization: adminProcedure
    .input(zOrganization.pick({publicMetadata: true}))
    .mutation(async ({ctx, input: update}) => {
      if (!ctx.viewer.orgId) {
        throw new TRPCError({code: 'BAD_REQUEST', message: 'orgId needed'})
      }
      const org = await clerkClient.organizations.updateOrganization(
        ctx.viewer.orgId,
        update,
      )
      return org
    }),
})

export async function getOrganizationOmitPrivateMeta(organizationId: string) {
  const org = await clerkClient.organizations.getOrganization({organizationId})
  return {
    ...R.pick(org, [
      'name',
      'slug',
      'imageUrl',
      'createdAt',
      'updatedAt',
      'membersCount',
    ]),
    id: zOrganization.shape.id.parse(org.id),
    publicMetadata: zOrganization.shape.publicMetadata.parse(
      org.publicMetadata,
    ),
  }
}

export async function getUserOmitPrivateMeta(userId: string) {
  const usr = await clerkClient.users.getUser(userId)
  return {
    ...R.pick(usr, ['id', 'imageUrl', 'createdAt', 'updatedAt']),
    id: zUser.shape.id.parse(usr.id),
    publicMetadata: zUser.shape.publicMetadata.parse(usr.publicMetadata),
    unsafeMetadata: zUser.shape.unsafeMetadata.parse(usr.publicMetadata),
  }
}
