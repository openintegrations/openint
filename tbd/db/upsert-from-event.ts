import {pgTable} from 'drizzle-orm/pg-core'

export function isValidDateString(str: string) {
  const date = new Date(str)
  const lenToCheck = Math.min(str.length, '2021-01-01T00:00:00'.length)
  return (
    !Number.isNaN(date.getTime()) &&
    date.toISOString().slice(0, lenToCheck) === str.slice(0, lenToCheck)
  )
}

export function inferTable(tableName: string, data: Record<string, unknown>) {
  return pgTable(tableName, (t) => {
    function inferCol(v: unknown) {
      if (typeof v === 'string') {
        return isValidDateString(v) ? t.timestamp() : t.text()
      }
      if (typeof v === 'number') {
        return Math.floor(v) === v ? t.integer() : t.doublePrecision()
      }
      if (typeof v === 'boolean') {
        return t.boolean()
      }
      if (v instanceof Date) {
        return t.timestamp()
      }
      return t.jsonb()
    }
    return Object.fromEntries(
      Object.entries(data).map(([k, v]) => [k, inferCol(v)]),
    )
  })
}
