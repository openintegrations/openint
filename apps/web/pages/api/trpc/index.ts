import type {NextApiRequest, NextApiResponse} from 'next'
import {renderTrpcPanel} from 'trpc-panel'
import {appRouter} from '@openint/api'

export const maxDuration = 60;

export default function handler(req: NextApiRequest, res: NextApiResponse) {
  // req.url is normally `/api/trpc` already which is the right place
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  res.status(200).send(renderTrpcPanel(appRouter, {url: req.url!}))
}
