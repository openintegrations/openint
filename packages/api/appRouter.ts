// not sure about directly depending on vdk from api, but anyways
import type {
  ZodOpenApiComponentsObject,
  ZodOpenApiPathsObject,
} from '@lilyrose2798/trpc-openapi/dist/generator'
import {generateOpenApiDocument} from '@lilyrose2798/trpc-openapi/dist/generator'
import {getServerUrl} from '@openint/app-config/constants'
import {flatRouter, outgoingWebhookEventMap} from '@openint/engine-backend'
import accountingRouter from '@openint/unified-accounting'
import atsRouter from '@openint/unified-ats'
import bankingRouter from '@openint/unified-banking'
import {crmRouter} from '@openint/unified-crm'
import eltRouter from '@openint/unified-etl'
import hrisRouter from '@openint/unified-hris'
import ptaRouter from '@openint/unified-pta'
import {salesEngagementRouter} from '@openint/unified-sales-engagement'
import {mapKeys, mapValues, publicProcedure, trpc, z} from '@openint/vdk'
import {authRouter} from './authRouter'

export const publicRouter = trpc.router({
  getOpenapiDocument: publicProcedure
    .meta({openapi: {method: 'GET', path: '/openapi.json'}})
    .input(z.void())
    .output(z.unknown())
    .query((): unknown => getOpenAPISpec()),
})

export const _appRouter = trpc.router({
  public: publicRouter,
  // Verticals
  salesEngagement: salesEngagementRouter,
  crm: crmRouter,
  banking: bankingRouter,
  accounting: accountingRouter,
  pta: ptaRouter,
  ats: atsRouter,
  hris: hrisRouter,
  etl: eltRouter,
})

export const appRouter = trpc.mergeRouters(flatRouter, authRouter, _appRouter)

export type AppRouter = typeof appRouter

function assertNoSlash(name: string) {
  if (name.includes('/')) {
    throw new Error(`Event name ${name} containing '/' is not supported`)
  }
  return name
}

export function oasWebhooksEventsMap(
  eMap: Record<string, {data: z.AnyZodObject}>,
) {
  const webhooks = mapValues(
    eMap,
    (_, name): ZodOpenApiPathsObject[string] => ({
      post: {
        requestBody: {
          content: {
            'application/json': {
              schema: {
                $ref: `#/components/schemas/webhooks.${assertNoSlash(name)}`,
              },
            },
          },
        },
        responses: {},
      },
    }),
  )
  type Schemas = NonNullable<ZodOpenApiComponentsObject['schemas']>
  const components = {
    schemas: mapKeys(
      mapValues(eMap, (shape, name): Schemas[string] =>
        z.object({...shape, name: z.literal(name), id: z.string().optional()}),
      ),
      (name) => `webhooks.${name}`,
    ),
  }
  return {webhooks, components}
}

export function getOpenAPISpec() {
  const {webhooks, components} = oasWebhooksEventsMap(outgoingWebhookEventMap)
  const oas = generateOpenApiDocument(appRouter, {
    openApiVersion: '3.1.0', // Want jsonschema
    title: 'OpenInt OpenAPI',
    version: '0.0.0',
    securitySchemes: {
      apikey: {
        type: 'apiKey',
        name: 'x-apikey',
        in: 'header',
      },
      resourceId: {
        type: 'apiKey',
        name: 'x-resource-id',
        in: 'header',
      },
    },
    baseUrl: getServerUrl(null) + '/api/v0',
    webhooks,
    components,
  })
  // Unfortunately trpc-openapi is missing bunch of options...
  oas.security = [{apikey: [], resourceId: []}]
  return oas
}

if (require.main === module) {
  console.log(JSON.stringify(getOpenAPISpec(), null, 2))
}
