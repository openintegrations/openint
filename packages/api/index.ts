import {appRouter} from './appRouter'
import {createRouterHandler} from './createRouterHandler'

export * from './appRouter'
export * from './createRouterHandler'
export * from './proxyHandler'

// TODO: Make me work
export function createAppHandler(
  opts: {
    endpoint?: `/${string}`
  } = {},
) {
  return createRouterHandler({...opts, router: appRouter})
}
