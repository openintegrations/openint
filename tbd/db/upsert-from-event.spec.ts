import {generateDrizzleJson, generateMigration} from 'drizzle-kit/api'
import {inferTable, isValidDateString} from './upsert-from-event'

test.each([
  ['2021-01-01T00:00:00Z', true],
  ['2021-01-01T00:00:00', false], // missing timezone, not valid as a result
  ['2021-01-01', true],
  ['2021-01', true],
  ['2021', true],
  ['12', false],
  ['202', false],
  ['2021-01-01T00:00:00+00:00', true],
  ['Sun, Nov 17, 5:00 PM', false],
  ['Sat Nov 16 2024 15:28:04 GMT-0800 (Pacific Standard Time)', false],
])('isValidDateString(%s) -> %o', (input, valid) => {
  expect(isValidDateString(input)).toEqual(valid)
})

test('infer', async () => {
  const table = inferTable('transaction', {
    id: '1',
    intId: 2,
    amount: 1.23,
    created_at: '2021-01-01T00:00:00Z',
    updated_at: new Date(),
    is_deleted: false,
    raw: {key: 'value'},
    array_value: [{key: 'value'}],
  })
  const migrations = await generateMigration(
    generateDrizzleJson({}),
    generateDrizzleJson({table}),
  )
  expect(migrations).toMatchSnapshot()
})
