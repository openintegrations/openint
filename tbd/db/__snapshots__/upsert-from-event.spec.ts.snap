// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`infer 1`] = `
[
  "CREATE TABLE IF NOT EXISTS "transaction" (
	"id" text,
	"intId" integer,
	"amount" double precision,
	"created_at" timestamp,
	"updated_at" timestamp,
	"is_deleted" boolean,
	"raw" jsonb,
	"array_value" jsonb
);
",
]
`;
